File: app/src/lib/stores/session.ts
import { writable, get } from 'svelte/store';
import { getInvoke } from '$lib/utils/tauri';
import { noteTitle } from './note';
import { currentWorkspace } from './workspace';
import { toast } from './toast';

export interface TrackerInfo {
  note_path: string;
  duration_minutes: number;
  opened_at: string;
}

interface SessionStoreState {
  trackerInfo: TrackerInfo | null;
  isLoading: boolean;
  error: string | null;
}

const defaultState: SessionStoreState = {
  trackerInfo: null,
  isLoading: false,
  error: null,
};

function createSessionStore() {
  const { subscribe, set, update } = writable<SessionStoreState>(defaultState);

  return {
    subscribe,

    // Start tracking a note
    startTracking: async (notePath: string) => {
      update((s) => ({ ...s, isLoading: true }));
      try {
        const invoke = await getInvoke();
        await invoke('start_tracking', { notePath });
        const info = await invoke<TrackerInfo | null>('get_tracker_info');
        update((s) => ({ ...s, trackerInfo: info, isLoading: false, error: null }));
      } catch (e) {
        const error = e instanceof Error ? e.message : String(e);
        console.error('[Session] Failed to start tracking:', error);
        toast.warning('Session tracking error');
        update((s) => ({ ...s, isLoading: false, error }));
      }
    },

    // Stop tracking and commit changes (when closing/switching notes)
    stopTracking: async (): Promise<TrackerInfo | null> => {
      try {
        const invoke = await getInvoke();
        const trackerInfo = await invoke<TrackerInfo | null>('stop_tracking');

        // Commit to git if we have valid tracking data
        if (trackerInfo && trackerInfo.note_path !== 'new-note') {
          const workspace = get(currentWorkspace);
          const title = get(noteTitle);

          if (workspace && title) {
            try {
              // Update metadata first
              await invoke('update_note_metadata', { notePath: trackerInfo.note_path });

              // Then commit
              const commitId = await invoke<string>('commit_session', {
                workspacePath: workspace.path,
                notePath: trackerInfo.note_path,
                title,
                durationMinutes: trackerInfo.duration_minutes,
              });
              console.log('[Session] Committed on file close:', commitId);
              toast.success(`Session saved (${formatDuration(trackerInfo.duration_minutes)})`);
            } catch (e) {
              console.error('[Session] Failed to commit on file close:', e);
            }
          }
        }

        set(defaultState);
        return trackerInfo;
      } catch (e) {
        const error = e instanceof Error ? e.message : String(e);
        console.error('[Session] Failed to stop tracking:', error);
        return null;
      }
    },

    // Reset store
    reset: () => {
      set(defaultState);
    },
  };
}

export const sessionStore = createSessionStore();

// Helper to format duration
export function formatDuration(minutes: number): string {
  if (minutes < 60) {
    return `${minutes}m`;
  }
  const hours = Math.floor(minutes / 60);
  const mins = minutes % 60;
  return mins > 0 ? `${hours}h ${mins}m` : `${hours}h`;
}

File: app/src-tauri/src/commands/session.rs
use crate::session::{TrackerInfo, TrackerManager};
use crate::storage::{load_metadata, save_metadata, NoteMeta};
use std::path::Path;
use std::sync::Mutex;
use tauri::State;

/// Wrapper for thread-safe tracker manager (Tauri managed state)
pub struct TrackerManagerState(pub Mutex<TrackerManager>);

impl TrackerManagerState {
    pub fn new() -> Self {
        Self(Mutex::new(TrackerManager::new()))
    }
}

impl Default for TrackerManagerState {
    fn default() -> Self {
        Self::new()
    }
}

/// Get current tracker info (duration, path)
#[tauri::command]
pub fn get_tracker_info(tracker_state: State<'_, TrackerManagerState>) -> Result<Option<TrackerInfo>, String> {
    let manager = tracker_state.0.lock()
        .map_err(|e| format!("Failed to acquire tracker lock: {}", e))?;
    Ok(manager.get_info())
}

/// Start tracking a note (called when opening a note)
#[tauri::command]
pub fn start_tracking(note_path: String, tracker_state: State<'_, TrackerManagerState>) -> Result<(), String> {
    let manager = tracker_state.0.lock()
        .map_err(|e| format!("Failed to acquire tracker lock: {}", e))?;
    manager.start_tracking(&note_path);
    Ok(())
}

/// Stop tracking and return duration info (called when closing/switching notes)
/// Returns the tracker data for use in commit
#[tauri::command]
pub fn stop_tracking(tracker_state: State<'_, TrackerManagerState>) -> Result<Option<TrackerInfo>, String> {
    let manager = tracker_state.0.lock()
        .map_err(|e| format!("Failed to acquire tracker lock: {}", e))?;
    let tracker = manager.stop_tracking();
    Ok(tracker.map(|t| {
        let duration = t.duration_minutes();
        TrackerInfo {
            note_path: t.note_path,
            duration_minutes: duration,
            opened_at: t.opened_at,
        }
    }))
}

/// Update metadata with the latest file info
#[tauri::command]
pub fn update_note_metadata(note_path: String) -> Result<(), String> {
    let path = Path::new(&note_path);

    // Load or create metadata
    let mut meta = load_metadata(path)
        .map_err(|e| format!("Failed to load metadata: {}", e))?
        .unwrap_or_else(|| NoteMeta::new(path));

    // Update file timestamp
    meta.update_file();

    // Save
    save_metadata(path, &meta).map_err(|e| format!("Failed to save metadata: {}", e))
}
